### ISSUE TITLE
crosspointer_transmute incorrectly fires on fn pointer transmute

### ISSUE THREAD

**Issue by SUPERCILEX** [rust-lang/rust-clippy#12067]
### Summary

AFAIK, this is the only way to call function pointers: https://doc.rust-lang.org/std/mem/fn.transmute.html#examples

### Lint Name

crosspointer_transmute

### Reproducer

I tried this code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f7eff6f54b49d52fb29de29380650f07

```rust
pub fn call(ptr: std::ptr::NonNull<fn()>) {
    unsafe { std::mem::transmute::<_, fn()>(ptr.as_ptr())() };
}
```

I saw this happen:

```
warning: transmute from a type (`*mut fn()`) to the type that it points to (`fn()`)
 --> src/lib.rs:2:14
  |
2 |     unsafe { std::mem::transmute::<_, fn()>(ptr.as_ptr())() };
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#crosspointer_transmute
  = note: `#[warn(clippy::crosspointer_transmute)]` on by default
```

I expected to see this happen:
No warning.

**Comment by Jarcho** [comment]
`fn()` is a function pointer type and is already non-null. `NonNull<fn()>` is a non-null pointer to a function pointer.

**Comment by SUPERCILEX** [comment]
Oh snap, I had to do a double take. You're right this is very dumb—I get to remove a bunch of unnecessary unsafe code though! Thanks!
